rust_src/src/eval.rs

```diff
@@ -948,4 +952,143 @@ fn run_hook_with_args_internal(
     }
 }
 
+enum LispFun {
+    SubrFun(LispSubrRef),
+    LambdaFun(LispObject),
+}
+
+enum LispFunError {
+    InvalidFun,
+    VoidFun,
+}
+
+fn resolve_fun(fun: LispObject) -> Result<LispFun, LispFunError> {
+    let original_fun = fun;
+
+    loop {
+        if original_fun.is_nil() {
+            return Err(LispFunError::VoidFun);
+        }
+
+        // Optimize for no indirection.
+        let fun = original_fun
+            .as_symbol()
+            .map_or_else(|| fun, |f| f.get_indirect_function());
+
+        if fun.is_nil() {
+            return Err(LispFunError::VoidFun);
+        }
+
+        if let Some(f) = fun.as_subr() {
+            return Ok(LispFun::SubrFun(f));
+        }
+
+        if unsafe { COMPILEDP(fun) || MODULE_FUNCTIONP(fun) } {
+            return Ok(LispFun::LambdaFun(fun));
+        }
+
+        if let Some(cons) = fun.as_cons() {
+            let funcar = cons.car();
+
+            if !funcar.is_symbol() {
+                return Err(LispFunError::InvalidFun);
+            }
+
```

Comment:
This does not read as nicely as the C code.

Why not use positive checks here?

```
if Qlambda or Qclosure {
    return Ok(LambdaFun);
}

if Qautoload {
   Fautoload_do_load(fun, original_fun, Qnil);
   unsafe { check_cons_list() };
   continue;
}
return Err(LispFunError::InvalidFun);
```

Reply:
Fixed